container/libcontainer/handler.go

```diff
@@ -143,7 +144,47 @@ func (h *Handler) GetStats() (*info.ContainerStats, error) {
 	return stats, nil
 }
 
-func processStatsFromProcs(rootFs string, cgroupPath string) (info.ProcessStats, error) {
+func processRootProcUlimits(rootFs string, rootPid int) []info.UlimitSpec {
+	filePath := path.Join(rootFs, "/proc", strconv.Itoa(rootPid), "limits")
+	out, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		klog.V(4).Infof("error while listing directory %q to read ulimits: %v", filePath, err)
+		return []info.UlimitSpec{}
+	}
+
+	limits := strings.Split(string(out), "\n")
+	ulimits := make([]info.UlimitSpec, 0, len(limits))
+	for _, lim := range limits {
+		// Skip any headers/footers
+		if strings.HasPrefix(lim, "Max") {
+			// Line format: Max open files            16384                16384                files
+			fields := regexp.MustCompile("[\\s]{2,}").Split(lim, -1)
+			name := strings.Replace(strings.ToLower(strings.TrimSpace(fields[0])), " ", "_", -1)
+
+			soft := strings.TrimSpace(fields[1])
+			soft_num, err := strconv.Atoi(soft)
+			if err != nil {
+				soft_num = -1
+			}
+
+			hard := strings.TrimSpace(fields[2])
+			hard_num, err := strconv.Atoi(hard)
+			if err != nil {
+				hard_num = -1
```

Comment:
I'm assuming -1 means unlimited?  How hard would it be to use the node-level limit if no container limit exists?

Reply:
@dashpole : yes, -1 means unlimited in general, with the exception of nice and priority. I think it might be better to drop the metric in that case, or only emit the soft/hard limit depending on which one has a valid value.

W.r.t. your second question: Do you mean host-level limit when you say node-level? The limits are inherited from dockerd (which in turn gets it from system settings or configurations), if they are not specified at the container level. https://docs.docker.com/engine/reference/commandline/dockerd/#default-ulimit-settings
So if you're concerned about what values to default to, that's already taken care of by Linux/dockerd.

Reply:
Yeah, I mean host-level.

So are you saying that the ulimit for the container already takes host-level limits into account?  I.E. if it is -1, then that means it is unlimited on the host?

Reply:
Yes. If it's not set at the container level, it inherits the value from dockerd, which in turn gets its ulimits from systemd, etc. In general, linux processes inherit ulimit values from their parent during a fork: http://danielzhou82.github.io/blog/2014/09/30/on-the-confusing-ulimit/  The process can set soft limit up to hard limit, but cannot change the hard limit, unless it is a privileged process.

Reply:
Thanks for explaining.